### PR TITLE
Use Cleaner to free digest native context

### PR DESCRIPTION
**Pull request by alon-sh**
Signed-off-by: Alon Shalev Housfater <alonsh@ca.ibm.com>

### CODE REVIEW

**Comment by alon-sh** [comment]
FYI @keithc-ca

File: closed/src/java.base/share/classes/sun/security/provider/NativeDigest.java [hunk 1]
```diff
@@ -55,10 +57,25 @@
...
+    final private static NativeCrypto nativeCrypto;
+    final private static Cleaner digestCleaner;
 
     static {
         nativeCrypto = NativeCrypto.getNativeCrypto();
+        digestCleaner = CleanerFactory.cleaner();
+    }
+
+    private static class DigestCleanerRunnable implements Runnable {
+        private long digestCtx;
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/sun/security/provider/NativeDigest.java]: Please apply the same fixes as were made for jdk8. Here, make the class and field `final`.

File: closed/src/java.base/share/classes/sun/security/provider/NativeDigest.java [hunk 2]
```diff
@@ -70,24 +87,24 @@
         this.digestLength = digestLength;
         this.algIndx = algIndx;
         this.context = nativeCrypto.DigestCreateContext(0, algIndx);
+        digestCleaner.register(this, new DigestCleanerRunnable(this.context));
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/sun/security/provider/NativeDigest.java]: The call to `register` should move after the validation check (also in `clone()`).

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 3]
```diff
@@ -86,12 +86,11 @@
     /*
      * Check whether native crypto is enabled with property.
      * By default, the native crypto is enabled and uses native library crypto.
-     * The native crypto for MessageDigest is disabled until
-     * https://github.com/eclipse/openj9/issues/5611 can be resolved.
+     * The property 'jdk.nativeDigest' is used to disable Native digest alone
      * The property 'jdk.nativeCrypto' is used to disable all native cryptos (Digest,
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java]: Please fix the comment: restore from commit 9ac2440.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by alon-sh** [comment]
@keithc-ca updated

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 4]
```diff
@@ -86,12 +86,11 @@
     /*
      * Check whether native crypto is enabled with property.
      * By default, the native crypto is enabled and uses native library crypto.
-     * The native crypto for MessageDigest is disabled until
-     * https://github.com/eclipse/openj9/issues/5611 can be resolved.
-     * The property 'jdk.nativeCrypto' is used to disable all native cryptos (Digest,
+     * The property 'jdk.nativeDigest' is used to disable Native digest alone
+     * and 'jdk.nativeCrypto' is used to disable all native cryptos (Digest,
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java]: Please remove trailing whitespace from this line.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by alon-sh** [comment]
updated

**Comment by keithc-ca** [comment]
Jenkins test extended plinux jdk12

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**